### PR TITLE
JDK-8276800 Fix table headers in NumericShaper.html

### PR DESCRIPTION
**Pull request by LudwikJaniuk**
This change introduces no visual difference, but improves a11y.
EDIT: There is a slight visual change regarding centering in a few cells.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276800](https://bugs.openjdk.java.net/browse/JDK-8276800): Fix table headers in NumericShaper.html

### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to 71f26cd7908a21448d30befc333110c3cbaaae1c
 * [Alexey Ivanov](https://openjdk.java.net/census#aivanov) (@aivanov-jdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6291/head:pull/6291` \
`$ git checkout pull/6291`

Update a local copy of the PR: \
`$ git checkout pull/6291` \
`$ git pull https://git.openjdk.java.net/jdk pull/6291/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6291`

View PR using the GUI difftool: \
`$ git pr show -t 6291`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6291.diff">https://git.openjdk.java.net/jdk/pull/6291.diff</a>

</details>

### CODE REVIEW

**Comment by aivanov-jdk** [comment]
Ludvig, could you please elaborate on how this change improves accessibility?

**Review by naotoj — APPROVED** [review]

**Comment by LudwikJaniuk** [comment]
@aivanov-jdk Gladly. [Doccheck](https://openjdk.java.net/projects/code-tools/doccheck/) flagged it, my understanding is that every row needs to have a unique row header to aid screen readers. The previous code had one cell work as header for two rows. 

One could perhaps look at [the standards](https://html.spec.whatwg.org/multipage/tables.html#attr-tdth-headers) and implement something even better, but this is what I came up with.

**Comment by aivanov-jdk** [comment]
> /integrate

I believe this is preliminary as our discussion isn't over yet.

**Comment by aivanov-jdk** [comment]
> @aivanov-jdk Gladly. [Doccheck](https://openjdk.java.net/projects/code-tools/doccheck/) flagged it, my understanding is that every row needs to have a unique row header to aid screen readers. The previous code had one cell work as header for two rows.
> 
> One could perhaps look at [the standards](https://html.spec.whatwg.org/multipage/tables.html#attr-tdth-headers) and implement something even better, but this is what I came up with.

This information should be added both to the JBS bug description and to this PR description, otherwise this comes out of nowhere.

I think this approach is wrong. However, to satisfy this requirement of Doccheck dummy column with numbers was added to many tables which _logically_ have no row headers.

You change the semantics of the table by moving the header to another column. The rendering should rather change too, otherwise visual representation becomes misleading.

To me, the table looks correct currently: *Arabic* Unicode range serves as the title for both rows. I wonder how it's read by a screen reader: the original version and the updated version. What we should strive for is for clarity of the read table rather than blindly following the rule where each row should have a title, provided that the title is available in this particular case.

@jonathan-gibbons, Can this error be marked as false positive in the Doccheck report? Can the tool be smarter?

**Comment by aivanov-jdk** [comment]
> To me, the table looks correct currently: _Arabic_ Unicode range serves as the title for both rows. I wonder how it's read by a screen reader: the original version and the updated version. What we should strive for is for clarity of the read table rather than blindly following the rule where each row should have a title, provided that the title is available in this particular case.

You may be right, the description of [`rowgroup`](https://html.spec.whatwg.org/dev/tables.html#attr-th-scope-rowgroup) keyword says such a header applies to all the cells in the group: _“The row group state means the header cell applies to all the remaining cells in the row group.”_

> You change the semantics of the table by moving the header to another column. The rendering should rather change too, otherwise visual representation becomes misleading.

Maybe a better way would be to make the second column a row column as you're suggesting but still keeping the existing row groups and changing the `scope` of the Tai Tham header cell to `rowgroup`: that is `<th scope="rowgroup">Tai Tham</th>`; and then for consistency with the rows above mark the second column with `<th scope="row">`.

Does it make any sense?

Is the third column a better candidate for the row header?

**Comment by LudwikJaniuk** [comment]
@aivanov-jdk I apologize, I haven't learned the etiquette around here yet.

**Comment by LudwikJaniuk** [comment]
I've pushed a change which "ticks the boxes without introducing so much change". I keep the first column as the row header, and simply specify it explicitly on the `td` elements. Doccheck is fine with it.

@aivanov-jdk If you thinks this works, I'd ask you to review and sponsor (if the review is necessary given the existing, outdated one).

File: src/java.desktop/share/classes/java/awt/font/NumericShaper.java [hunk 1]
```diff
@@ -112,25 +112,22 @@
...
- *   </tr>
+ *     <td headers="rowheader1 colheader3">{@link NumericShaper#EASTERN_ARABIC NumericShaper.EASTERN_ARABIC}
  *   <tr>
- *     <td>{@link NumericShaper.Range#ARABIC}
+ *     <td headers="rowheader1 colheader2">{@link NumericShaper.Range#ARABIC}
  *     <br>
  *     {@link NumericShaper.Range#EASTERN_ARABIC}
- *     <td>{@link NumericShaper.Range#EASTERN_ARABIC}
- * </tbody>
- * <tbody>
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/java/awt/font/NumericShaper.java:133]: You need to preserve both <tbody> to create have two rowgroups.

File: src/java.desktop/share/classes/java/awt/font/NumericShaper.java [hunk 2]
```diff
@@ -112,25 +112,22 @@
...
+ *     <td headers="rowheader1 colheader3">{@link NumericShaper.Range#EASTERN_ARABIC}
  *   <tr>
  *     <th scope="row">Tai Tham
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/java/awt/font/NumericShaper.java:132]: Does it make sense to use `<th scope="rowgroup">` for consistency?

File: src/java.desktop/share/classes/java/awt/font/NumericShaper.java [hunk 3]
```diff
@@ -112,25 +112,22 @@
  * <caption>NumericShaper constants precedence</caption>
  * <thead>
  *   <tr>
- *     <th scope="col">Unicode Range
- *     <th scope="col">{@code NumericShaper} Constants
- *     <th scope="col">Precedence
+ *     <th scope="col" id="colheader1">Unicode Range
+ *     <th scope="col" id="colheader2">{@code NumericShaper} Constants
+ *     <th scope="col" id="colheader3">Precedence
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/java/awt/font/NumericShaper.java:117]: Shall we use more descriptive ids?
```suggestion
 *     <th scope="col" id="nsUnicodeRangeHeader">Unicode Range
 *     <th scope="col" id="nsConstantsHeader">{@code NumericShaper} Constants
 *     <th scope="col" id="nsPrecedence">Precedence
```

**Review by aivanov-jdk — COMMENTED** [review]
I'm fine with the 2nd version yet I'm pretty DocChecker won't be happy unless you create additional row header with `<th>` element.

Shall the third column of the table be converted to `<th>` to compensate for that?

**Comment by LudwikJaniuk** [comment]
@aivanov-jdk Your comment about `<tbody>` made me dig deeper, and I found a more in-depth [article on MDN](https://developer.mozilla.org/en-US/docs/Learn/HTML/Tables/Advanced) with guidance on this specific kind of tables. 
Quote:

> scope has two more possible values — colgroup and rowgroup. these are used for headings that sit over the top of multiple columns or rows. If you look back at the "Items Sold August 2016" table at the start of this section of the article, you'll see that the "Clothes" cell sits above the "Trousers", "Skirts", and "Dresses" cells. All of these cells should be marked up as headers (<th>), but "Clothes" is a heading that sits over the top and defines the other three subheadings. "Clothes" therefore should get an attribute of scope="colgroup", whereas the others would get an attribute of scope="col".

File: src/java.desktop/share/classes/java/awt/font/NumericShaper.java (hunk 1 above)
Comment (LudwikJaniuk) [reply src/java.desktop/share/classes/java/awt/font/NumericShaper.java:133]: Could you say where this is specified? See my comment below.

File: src/java.desktop/share/classes/java/awt/font/NumericShaper.java (hunk 3 above)
Comment (LudwikJaniuk) [reply src/java.desktop/share/classes/java/awt/font/NumericShaper.java:117]: Not relevant anymore given last change

File: src/java.desktop/share/classes/java/awt/font/NumericShaper.java (hunk 2 above)
Comment (LudwikJaniuk) [reply src/java.desktop/share/classes/java/awt/font/NumericShaper.java]: Not relevant anymore given last change

**Comment by aivanov-jdk** [comment]
> Concerning `<tbody>`, quote (emphasis mine):
> 
> > As your tables get a bit more complex in structure, it is useful to give them more structural definition. One clear way to do this is by using `<thead>`, `<tfoot>`, and `<tbody>`, which allow you to mark up a header, footer, and body section for the table.
> 
> > **These elements don't make the table any more accessible to screenreader users**, and don't result in any visual enhancement on their own. They are however very useful for styling and layout — acting as useful hooks for adding CSS to your table.
> 
> Based on this, I don't see a foundation for having several tbodies. Am I missing something?

The definition of [`rowgroup` keyword](https://html.spec.whatwg.org/dev/tables.html#attr-th-scope-rowgroup) is as follows:
> The row group state means the header cell applies to all the remaining cells in the row group. A `th` element's `scope` attribute must not be in the row group state if the element is not anchored in a row group.

As I read the definition, having `<th scope="rowgroup">` requires that the corresponding `th` element is in a row group (`tbody`) and such _a heading applies to all the cells in the row group_. Thus, without creating the second row group the cell `<th scope="rowgroup">Arabic` from the first row applies to all the subsequent cells in the table including `<th scope="rowgroup">Tai Tham`.

[The following example](https://html.spec.whatwg.org/dev/tables.html#the-th-element:attr-th-scope-6) on that page discusses the same situation: a heading cell with `scope="rowgroup"` applies to all the cells in a row group defined by `tbody`. Our situation is similar yet the second row group contains only one row.

It's why I think the table must have *two* row groups: for _Arabic_ and for _Tai Tham_, even though it doesn't change the visual presentation.

File: src/java.desktop/share/classes/java/awt/font/NumericShaper.java (hunk 3 above)
Comment (aivanov-jdk) [reply src/java.desktop/share/classes/java/awt/font/NumericShaper.java:117]: Why is it not relevant? You're still using the ids?
I see you've removed the `header` attribute but I still `id` attributes. Will you be removing the id attributes too?

**Review by aivanov-jdk — APPROVED** [review]
Not ideal, but I guess it's the best we can do.

**Comment by LudwikJaniuk** [comment]
There were arguments in our discussion earlier about letting the rendering reflect the HTML semantics, so I'm inclined to leave it like this.

**Pull request closed without merging**